### PR TITLE
Market decentralized - provider counter draft proposal

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #363
resolves: #376

### CODE REVIEW

File: core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql [hunk 1]
```diff
@@ -72,7 +72,6 @@ CREATE TABLE market_negotiation(
     offer_id VARCHAR(97) NOT NULL,
     demand_id VARCHAR(97) NOT NULL,
 
-    identity_id VARCHAR(20) NOT NULL,
```
Comment (tworec) [inline core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql:75]: 👍

File: core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql [hunk 2]
```diff
@@ -82,6 +81,7 @@ CREATE TABLE market_negotiation(
 CREATE TABLE market_proposal(
     id VARCHAR(100) NOT NULL PRIMARY KEY,
     prev_proposal_id VARCHAR(100),
+    issuer INTEFER NOT NULL,
```
Comment (tworec) [inline core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql:84]:
```suggestion
    issuer INTEGER NOT NULL,
```

File: core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql (hunk 1 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql:75]: I knew, you will be happy :P

File: core/market/decentralized/tests/test_negotiations.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        .subscribe_demand(&sample_demand(), &identity1)
+        .await?;
+    let offer_id = market2.subscribe_offer(&sample_offer(), &identity2).await?;
+
+    // Expect events generated on requestor market.
+    let events = market1
+        .requestor_engine
+        .query_events(&demand_id, 1.2, Some(5))
+        .await?;
+    let proposal0 = requestor::expect_proposal(events)?;
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:35]: would be good to assert `proposal0` props below eg. status Initial

File: core/market/decentralized/src/negotiation/common.rs [hunk 4]
```diff
@@ -1,28 +1,193 @@
...
+        let prev_proposal = self
+            .get_proposal(prev_proposal_id)
+            .await
+            .map_err(|e| ProposalError::from(&subscription_id, e))?;
+
+        log::debug!(
+            "Expected subscription id [{}], but found [{}].",
+            subscription_id,
+            prev_proposal.negotiation.subscription_id
+        );
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/common.rs:48]: I was going to remove these logs. Thanks for noticing.

File: core/market/decentralized/src/db/models/proposal.rs [hunk 5]
```diff
@@ -149,6 +156,25 @@ impl Proposal {
         }
     }
 
+    pub fn from_counter(&self, proposal_id: ProposalId, demand: ModelDemand) -> Proposal {
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/models/proposal.rs]: done

File: core/market/decentralized/src/db/models/proposal.rs [hunk 6]
```diff
@@ -163,6 +189,7 @@ impl Proposal {
 
         self.body.prev_proposal_id = Some(self.body.id.clone());
         self.body.id = proposal_id;
+        self.body.issuer = IssuerType::Us;
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/models/proposal.rs]: done

File: core/market/decentralized/src/negotiation/common.rs [hunk 7]
```diff
@@ -1,28 +1,193 @@
...
+            }
+        };
+
+        let owner_id = NodeId::from_str(&caller)
+            .map_err(|e| RemoteProposalError::Unexpected(e.to_string()))?;
+
+        let proposal_id = msg.proposal.proposal_id.clone();
+        let demand = msg
+            .proposal
+            .into(owner_id, prev_proposal.negotiation.demand_id.clone());
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/negotiation/common.rs]: done

**Pull request merged**